### PR TITLE
[DPE-3730][DPE-3729] Add Karapace client and auth handling

### PR DESCRIPTION
**Pull request by zmraul**
Adds `karapace_client` interface and handling of Auth for Karapace

-------------------------------------------
### Auth handling
Auth is handled on `managers/auth.py`. `authfile.json` from karapace is split into two sections:
- `users`: credentials for each user
- `permissions` list of ACLs, each entry on this list will be an ACL applied to a specific user

An example of `authfile.json` can be [found upstream](https://github.com/Aiven-Open/karapace?tab=readme-ov-file#example-of-complete-authorization-file). Authfile is loaded at the beginning of each event execution. Then, an internal mapping of auth is created inside `KarapaceAuth` class, on `self.auth_dict` attribute. Changes made to this internal state of auth are not persisted to the file until a specific method is invoked.

#### `KarapaceAuth` API:
- `add_user()`: adds the credentials for a user. It can replace credentials for an existing user or skip if user already exists
- `add_acl()`: gives permissions to a user, for a subject based on role (`admin` or `user`)
- `remove_user()`: removes user and acls associated with it
- `write_authfile()`: persists changes into the authfile

-------------------------------------------
### `karapace_client` interface
The interface inherits from `data_interfaces` classes. It can be found under `relations/karapace.py` The interface follows same pattern as `kafka_client` interface. Available roles are just two at the moment: `admin` and `user`.

Since a user doesn't really need to create anything, just generate permissions to access a schema created by an admin, the interface task is to allow access to a subject (a key linked to a schema). More info on [subject naming strategies and schemas](https://developer.confluent.io/courses/schema-registry/schema-subjects/).

The requirer requests access to a subject, that is handled on the provider, `SubjectRequestedEvent`. Depending on `user-roles`, permissions are assigned to a user on a subject. Then, `SubjectAllowedEvent` will trigger on the requirer to inform that the permissions have been created. At the moment the permissions are not dynamic, and set based on roles.

### CODE REVIEW

File: src/managers/auth.py [hunk 1]
```diff
@@ -25,85 +27,148 @@ class Acl:
...
+        """Returns the dict with an credentials representation."""
+        return {
+            "username": self.username,
+            "algorithm": self.algorithm,
+            "salt": self.salt,
+            "password_hash": self.password_hash,
+        }
+
+
+class AuthDictEntry(TypedDict):
```
Comment (marcoppenheimer) [inline src/managers/auth.py:62]: **question/todo:** Is there a specific reason in this being a TypedDict and not a dataclass?
If not, let's make it a dataclass to be consistent.

Comment (marcoppenheimer) [reply src/managers/auth.py:62]: **praise:** These mini types are really good. Meaningful, domain specific 👍🏾

File: src/managers/auth.py [hunk 2]
```diff
@@ -25,85 +27,148 @@ class Acl:
...
+        authfile = json.loads("\n".join(raw_file))
+
+        users = [UserCredentials(**user) for user in authfile["users"]]
+        acls = [Acl(**acl) for acl in authfile["permissions"]]
+
+        for user in users:
+            self.auth_dict[user.username] = AuthDictEntry(
+                credentials=user,
+                acls=[acl for acl in acls if acl.username == user.username],
+            )
```
Comment (marcoppenheimer) [inline src/managers/auth.py:108]: **thought:** (non-blocking) I get the feeling that this structure is overly complicated. I don't have a better idea off the top of my head so it's definitely fine to keep, but I wonder if the 'gain' of being able to reference `auth_dict[username]` is worth the JSON/Dict transformations.

File: src/managers/auth.py [hunk 3]
```diff
@@ -25,85 +27,148 @@ class Acl:
...
+
+        user_creds = UserCredentials(
+            **json.loads(self.workload.mkpasswd(username=username, password=password))
+        )
+        # If username doesn't exist we need to initialize the internal dictionary.
+        # Maybe default dict is a more elegant solution here.
+        if username not in self.auth_dict:
+            self.auth_dict[username] = AuthDictEntry(credentials=user_creds, acls=[])
+        else:
+            self.auth_dict[username]["credentials"] = user_creds
```
Comment (marcoppenheimer) [inline src/managers/auth.py:124]: **suggestion:** I think here you can use dict merging, creating an `AuthDictEntry` either way, and then merging it to `auth_dict`.
You might need to make `AuthDictEntry.acls` optional here, but you'd end up with something like:

```python
self.auth_dict = {user: {"credentials": existing_creds, "acls": existing_acls}}
new_entry = {user: {"credentials": new_creds}}

self.auth_dict = self.auth_dict | new_entry

print(self.auth_dict)

>>> {user: {"credentials": new_creds, "acls": existing_acls}}
```

File: src/managers/auth.py [hunk 4]
```diff
@@ -25,85 +27,148 @@ class Acl:
...
+class UserCredentials:
+    """Object for representing credentials of a Karapace user."""
+
+    username: str
+    algorithm: Algorithm
+    salt: str
+    password_hash: str
+
+    @property
+    def creds_repr(self) -> dict[str, str]:
```
Comment (marcoppenheimer) [inline src/managers/auth.py:52]: **todo:** You can remove this method, and use `dataclasses.asdict` wherever you need it

```python
from dataclasses import dataclass, asdict

@dataclass
class X:
    foo: str

x = X(foo="hello")
print(asdict(x))

>>> {'foo': 'hello'}
```

File: src/relations/karapace.py
Comment (marcoppenheimer) [inline src/relations/karapace.py:1]: **praise:** General point, but I REALLY like having relations be part of the core code, and not hidden in libs. I wish we could do that with all the 'requires/provides' objects for the other charms.

**Review by marcoppenheimer — CHANGES_REQUESTED** [review]
I left a lot of comments, but overall absolutely fantastic job. Pleasure to read through, really clean, well structured and thought out, excellent job!

Once questions are addressed, I'm happy to approve as is if you don't fancy any of the thoughts/suggestions raised.

File: src/events/password_actions.py [hunk 5]
```diff
@@ -64,12 +64,12 @@ def _set_password_action(self, event: ActionEvent) -> None:
...
 
-        try:
-            self.charm.auth_manager.add_user(username=username, password=new_password)
-        except Exception as e:
-            logger.error(str(e))
-            event.fail(f"unable to set password for {username}")
-            return
+        self.charm.auth_manager.add_user(username=username, password=new_password, replace=True)
+        self.charm.auth_manager.add_acl(username=username, role="admin")
+        self.charm.auth_manager.write_authfile()
```
[the remark this answers is not in the document]
Comment (zmraul) [reply src/events/password_actions.py:69]: The creation of passwords and writing down the file here are static utils. They don't depend on the service being up. I could keep the general `Exception` catch as a safeguard though.

File: src/managers/auth.py (hunk 4 above)
Comment (zmraul) [reply src/managers/auth.py:52]: TIL, thanks!

File: src/managers/auth.py (hunk 2 above)
Comment (zmraul) [reply src/managers/auth.py:112]: The problem addressed here is that we don't have a way to link a single username to both credentials and acls with the original json format. If I don't merge them into the internal dictionary here, then I would need to always iterate over all acls to find the ones that have `username == <my-username>`.

With the current approach, it's only done once when loading the file.

**Pull request merged**